### ISSUE TITLE
ComboBox popup drifts when the screen reshapes

### ISSUE THREAD

**Issue by samreid** [phetsims/sun#237]
Discovered during #229 ComboBox popup drifts when the screen reshapes.

![image](https://cloud.githubusercontent.com/assets/679486/14695791/bc335558-0732-11e6-942b-02f607c62766.png)

**Comment by pixelzoom** [comment]
Is this specific to CCK? Are you relocating the combo box, but not relocating the popup's parent?

**Comment by pixelzoom** [comment]
No such drift problem with combo boxes in Beer's Law Lab, so I'm guessing this is a CCK problem.

**Comment by samreid** [comment]
In CCK, the combo box parent is the ScreenView.  This is a problem for all sims that do not move the combo box parent (including but not limited to CCK).  Are there good reasons not to update the popup location in ComboBox.js when the visible screen bounds or layout bounds change?

**Comment by pixelzoom** [comment]
> Are there good reasons not to update the popup location in ComboBox.js when the visible screen bounds or layout bounds change?

Yes. It's unnecessary.  Instead of using the ScreenView as the listParent, just add an addition Node to the ScreenView, and use that as the listParent.

**Comment by samreid** [comment]
> Yes. It's unnecessary. Instead of using the ScreenView as the listParent, just add an addition Node to the ScreenView, and use that as the listParent.

But I would need to keep that additional Node position synchronized with the position of the ComboBox, right?

**Comment by pixelzoom** [comment]
> But I would need to keep that additional Node position synchronized with the position of the ComboBox, right?

No.  Changing the size of the window will have no affect on the listParent position.

```
var listParent = new Node();
myScreenView.addChild( listParent );
var myComboBox = new ComboBox( ..., listParent, ...)
```

**Comment by samreid** [comment]
@pixelzoom suggested dismissing the popup when the combobox changes location.  I can coordinate with @jonathanolson to find out the best way to determine whether the combobox has changed location.

**Comment by samreid** [comment]
@pixelzoom said we could add an API to ComboBox to programmatically dismiss the popup, though the more general solution may be preferable.

**Comment by samreid** [comment]
@jonathanolson is there a nice scenery way to get a callback when a node's global position has changed?  Would that work for this case?

**Comment by samreid** [comment]
@jonathanolson said:

> See if TransformTracker is sufficient?

**Comment by samreid** [comment]
It looks like TransformTracker requires the trail, so we could not wire this up automatically in ComboBox's constructor, we would have to wait until it is added into the scene graph.  Alternatively, we could pass in the screen view bounds property as an option to combo box, and when that changes it could collapse the combo box.  @pixelzoom and @jonathanolson what do you think?  Or should we just punt on this issue since it may be rare in practice?

**Comment by pixelzoom** [comment]
I would punt for now.  But perhaps we should have a general issue somewhere (joist?) that provides a more general API that could be used by all popups (combo box, menus, etc.)

**Comment by ariel-phet** [comment]
Marking for dev meeting as I would be curious how much work this would be (perhaps @pixelzoom suggestion of a general API that could allow us to dismiss all pop ups)

**Comment by pixelzoom** [comment]
Yes, ComboBox should eventually use the same "pop up" mechanism as menu, dialogs, etc.  But since this issue is reportedly specific to CCK, I don't believe it requires and changes to ComboBox.

**Comment by jonathanolson** [comment]
This was an issue in masses-and-springs, where I noticed two potential issues:

1. When the ComboBox was moved, the popped-up content didn't shift to compensate.
2. If the list parent was transformed before things popped up, the transform didn't seem to compensate.

**Comment by Denz1994** [comment]
This drifting behavior is also present in projectile-motion.

> Alternatively, we could pass in the screen view bounds property as an option to combo box, and when that changes it could collapse the combo box.

This sounds reasonable, because changing screen bounds would require a user input unrelated to the ComboBox. Any other similar action currently closes the pop-up so this seems like a natural extension of that behavior.

**Comment by jessegreenberg** [comment]
That would be a good short term workaround until a more general pop up system is ready.

**Comment by samreid** [comment]
Is there a more general way to listen for the global bounds of the node, and automatically remove the popup when the global bounds changes?

**Comment by jessegreenberg** [comment]
How about
```js
    listParent.on( 'bounds', hideList );
```
Are we trying to avoid usage of Events?

**Comment by Denz1994** [comment]
Recalling from memory: This was assigned to me to apply a listParentNode to the comboBox in Masses and Springs as suggested in the above comment https://github.com/phetsims/sun/issues/237#issuecomment-395500658.

This fix was applied to MAS and isn't an issue in that sim. I checked for usages of `new ComboxBox` and `Combox.call(` and this pattern is being used in those cases as well.

I pushed a fix to CCK-BBS, so I'll close this issue.
